lib/internal/modules/esm/translators.js

```diff
@@ -158,12 +162,27 @@ translators.set('wasm', async function(url) {
   }
 
   const imports =
-      WebAssembly.Module.imports(compiled).map(({ module }) => module);
+    WebAssembly.Module.imports(compiled).map(({ module }) => module);
   const exports = WebAssembly.Module.exports(compiled).map(({ name }) => name);
+  const usesWasi = wasiEnabled && imports.includes('wasi_unstable');
 
-  return createDynamicModule(imports, exports, url, (reflect) => {
-    const { exports } = new WebAssembly.Instance(compiled, reflect.imports);
-    for (const expt of Object.keys(exports))
-      reflect.exports[expt].set(exports[expt]);
-  });
+  return createDynamicModule(
+    usesWasi ? imports.filter((m) => m !== 'wasi_unstable') : imports,
+    exports, url,
+    (reflect) => {
+      let wasi;
+      if (usesWasi) {
+        wasi = new WASI(WASI.defaultConfig);
+        reflect.imports.wasi_unstable = wasi.exports;
+      }
+
+      const { exports } = new WebAssembly.Instance(compiled, reflect.imports);
+      for (const expt of Object.keys(exports))
+        reflect.exports[expt].set(exports[expt]);
+
+      if (usesWasi) {
+        wasi.setMemory(exports.memory);
```

Comment:
In this case, you're adding logic to the module loader which is WASI-specific and not on a standards track. We're beginning to discuss this need in https://github.com/WebAssembly/esm-integration/issues/32. This patch
makes the need for the logic clear in a way I didn't understand before, so thanks for that.

Reply:
In an ideal world I don't think this would exist. I'm hoping that some sort of first-class memory view type will be added which can be passed to WASI calls instead of the loader needing to mess with scoping weird out of band memory exports.